### ISSUE TITLE
1.2.4 breaks connections with openvpn tunnel interface

### ISSUE THREAD

**Issue by justinkb** [arvidn/libtorrent#4388]
libtorrent version (or branch): 1.2.4

platform/architecture: armv7h

compiler and compiler version: GCC 9.2.0

I have set up qBittorrent (which uses libtorrent) to only use the openvpn-client.service-created interface (tun0 in my case). In < 1.2.4 this worked fine since time immemorial, however when I update to 1.2.4 this causes AFAICT literally nothing to connect anymore (can't get peers via DHT, can't connect to trackers, peers, anything)

This is on archlinuxarm running on a Raspberry Pi 4. I looked through the open and closed issues extensively and couldn't quite find anything that matches my issue exactly, apologies if I overlooked something

**Comment by arvidn** [comment]
I think this is the same issue raised by https://github.com/arvidn/libtorrent/issues/4329 and https://github.com/arvidn/libtorrent/issues/4325

This is the main issue I've focused on fixing for a 1.2.5 release. In fact, I'm very close to cut that release. Any help in verifying the issue is fixed would be highly appreciated. If you can build and test libtorrent out of the `RC_1_2` branch, please do!

The one remaining issue is recovering the DHT connection when turning a VPN on and off.

**Comment by justinkb** [comment]
Right now I'm not able to access that Pi, but I will definitely test that ASAP once I am able, later today

**Comment by Ryrynz** [comment]
@justinkb Waiting on you mate..

**Comment by NotKyros** [comment]
@arvidn Not OP, but I follow the issue tracker on here. I pulled the RC_1_2 branch and tested it being bound to tun0. Everything works perfectly fine as I can tell. I didn't test DHT but normal tracker operations are successful and I was able to download and seed torrents.

**Comment by arvidn** [comment]
I believe this has been fixed in the 1.2.5 release

**Comment by justinkb** [comment]
I'm seeing the exact same behavior with 1.2.5 as with 1.2.4, reverting to 1.2.3 fixes the issue.

Magnet downloads get stuck on "Downloading metadata" and .torrent files uploaded to the qBittorrent webserver get stuck on "Stalled". No connections with other peers are being established at all.

**Comment by arvidn** [comment]
I think the most helpful would be to provide logs, with debug logging enabled.
Any error message would help, really. Do you see messages saying it successfully opened listen sockets?

**Comment by justinkb** [comment]
Output with both 1.2.3 and 1.2.5 is identical (apart from timestamps obviously):

```
(N) 2020-03-10T16:10:29 - qBittorrent v4.2.1 started
(N) 2020-03-10T16:10:29 - Using config directory: /var/lib/kodi/.config/qBittorrent/
(I) 2020-03-10T16:10:29 - Trying to listen on: tun0:9001
(N) 2020-03-10T16:10:29 - Peer ID: -qB4210-
(N) 2020-03-10T16:10:29 - HTTP User-Agent is 'qBittorrent/4.2.1'
(I) 2020-03-10T16:10:29 - DHT support [ON]
(I) 2020-03-10T16:10:29 - Local Peer Discovery support [ON]
(I) 2020-03-10T16:10:29 - PeX support [ON]
(I) 2020-03-10T16:10:29 - Anonymous mode [ON]
(I) 2020-03-10T16:10:29 - Encryption support [ON]
(I) 2020-03-10T16:10:29 - GeoIP database loaded. Type: GeoLite2-Country. Build time: Tue Dec 24 18:40:14 2019.
(N) 2020-03-10T16:10:29 - Using built-in Web UI.
(N) 2020-03-10T16:10:29 - Web UI translation for selected locale (en_US) has been successfully loaded.
(N) 2020-03-10T16:10:29 - Web UI: Now listening on IP: *, port: 8880
(I) 2020-03-10T16:10:29 - Successfully listening on IP: 10.8.1.11, port: TCP/9001
(I) 2020-03-10T16:10:29 - Successfully listening on IP: 10.8.1.11, port: UDP/9001
(I) 2020-03-10T16:10:29 - Successfully listening on IP: fe80::8c69:ea3d:ad75:ded%tun0, port: TCP/9001
(I) 2020-03-10T16:10:29 - Successfully listening on IP: fe80::8c69:ea3d:ad75:ded%tun0, port: UDP/9001
(N) 2020-03-10T16:10:29 - 'ubuntu-19.10-desktop-amd64.iso' restored.
(W) 2020-03-10T16:10:29 - Couldn't download GeoIP database file. Reason: The remote host name was not found (invalid hostname)
(I) 2020-03-10T16:10:29 - Detected external IP: my.vpn.ip.address
```

This is from qBittorrent logs, not sure if there is a way to increase how detailed the output is

**Comment by justinkb** [comment]
I'll try setting up a toolchain targetting archlinuxarm for my pi so bisecting this issue doesn't take forever. Meanwhile this issue should be reopened since the problem is definitely not fixed

**Comment by arvidn** [comment]
I would rather like to know why it doesn't work, than to know which change broke it. Probably not binding to 0.0.0.0 is what caused this issue. Although, none of the evidence in this thread suggests that though. If you have the capacity to build, please enable logging instead of bisecting.

**Comment by justinkb** [comment]
I posted the qBittorrent logs above, how would I go about producing libtorrent debug log output?

**Comment by arvidn** [comment]
My understanding is that qBT discards some alerts (most notably tracker alerts), which may require some patching of the alert handler in qBT. I'm not sure what would need to be done there.

To get libtorrent to print more verbose logs, set the `settings_pack::alert_mask` setting to a combination of these [flags](https://libtorrent.org/reference-Alerts.html#error_notification).

Probably: 1 + 4 + 32 + 64 + 1024 + 8192 + 16384 + 131072 + 524288 = 681061

**Comment by justinkb** [comment]
Patched it with lt::alert::all_categories to be safe, building now

**Comment by arvidn** [comment]
then you might have to increase `alert_queue_size` as well. To, say, 10000. Otherwise the rate of messages can exceed the queue size and get dropped.

**Comment by justinkb** [comment]
Thanks for the heads up, looks like qBittorrent by default sets it to int_max divided by 2, which should be enough, I'll let you know once I have some results

**Comment by justinkb** [comment]
after a reinstall of my system, with roughly the same configuration I can no longer reproduce the issue

**Comment by justinkb** [comment]
update on this: turns out the above was in fact with libtorrent 1.2.3, so the issue does still exist in 1.2.4 and beyond.

does seem like a duplicate of https://github.com/arvidn/libtorrent/issues/4412 as I'm also using an isolated vpn set up for torrenting

**Comment by justinkb** [comment]
So I've built qbittorrent on my new set up with the alertMask set to full, doesn't really seem to change anything/much in terms of logging output. I noticed tracker_notifications were already turned on by default for qbittorrent. Anyway, here's the output:

```
(N) 2020-05-14T18:22:52 - qBittorrent v4.2.5 started
(N) 2020-05-14T18:22:52 - Using config directory: /var/lib/kodi/.config/qBittorrent/
(I) 2020-05-14T18:22:52 - Trying to listen on: 0.0.0.0:9001,[::]:9001
(N) 2020-05-14T18:22:52 - Peer ID: -qB4250-
(N) 2020-05-14T18:22:52 - HTTP User-Agent is 'qBittorrent/4.2.5'
(I) 2020-05-14T18:22:52 - DHT support [ON]
(I) 2020-05-14T18:22:52 - Local Peer Discovery support [ON]
(I) 2020-05-14T18:22:52 - PeX support [ON]
(I) 2020-05-14T18:22:52 - Anonymous mode [ON]
(I) 2020-05-14T18:22:52 - Encryption support [FORCED]
(N) 2020-05-14T18:22:52 - Using built-in Web UI.
(W) 2020-05-14T18:22:52 - Couldn't load Web UI translation for selected locale (C).
(N) 2020-05-14T18:22:52 - Web UI: Now listening on IP: *, port: 8880
(I) 2020-05-14T18:22:52 - Successfully listening on IP: 127.0.0.1, port: TCP/9001
(I) 2020-05-14T18:22:52 - Successfully listening on IP: 127.0.0.1, port: UDP/9001
(I) 2020-05-14T18:22:52 - Successfully listening on IP: 10.7.3.3, port: TCP/9001
(I) 2020-05-14T18:22:52 - Successfully listening on IP: 10.7.3.3, port: UDP/9001
(I) 2020-05-14T18:22:52 - Successfully listening on IP: ::1, port: TCP/9001
(I) 2020-05-14T18:22:52 - Successfully listening on IP: ::1, port: UDP/9001
(I) 2020-05-14T18:22:52 - Successfully listening on IP: fe80::520:6896:56ee:459a%tun0, port: TCP/9001
(I) 2020-05-14T18:22:52 - Successfully listening on IP: fe80::520:6896:56ee:459a%tun0, port: UDP/9001
(N) 2020-05-14T18:26:19 - WebAPI login success. IP: ::ffff:127.0.0.1
```

I can't connect to any tracker, nor does DHT ever get bootstrapped. Inside the namespaced openvpn network I have isolated qBittorrent in, every other network functionality I've tried works fine and correctly shows I'm connecting through the vpn, so that can't be the problem

**Comment by arvidn** [comment]
@justinkb which version of libtorrent is this using? do you know?

**Comment by justinkb** [comment]
Built from this exact commit
https://github.com/arvidn/libtorrent/commit/a9968916ca82366f1c236af59aaecb9bc94ffe73

**Comment by arvidn** [comment]
@justinkb thanks! yeah, there hasn't been any more attempts after that commit to fix this issue.

When you say you can't connect to any tracker, that includes HTTP trackers too, right? not just UDP ones (I would expect that if the DHT doesn't work, UDP trackers wouldn't either).

To narrow down what's going on, would you mind testing with setting the listen interface to "10.7.3.3"? (which I assume is your tunnel out to the VPN).

I suspect you don't have a default route for 10.7.3.3 as the source address (or for the device that IP is associated with) and that confuses libtorrent.

The fundamental problem is that libtorrent wants to announce both your IPv4 and IPv6 address to the tracker, but not 127.0.0.1 nor ::1, so it has a heuristic to figure out which addresses can be used as source IP and reach the tracker.

**Comment by justinkb** [comment]
Progress! If I start the qbittorrent process from an interactive shell while inside the network namespace it does work perfectly fine from the looks of it. Both UDP and HTTP trackers seem to work, as well as DHT. Now to figure out why it doesn't work when using NetworkNamespacePath in the systemd service [Service] section.

**Comment by arvidn** [comment]
@justinkb if it's not too much trouble, could you post the network interface configuration and the routing table from *within* your container?

on linux:

`ifconfig` and `netstat -rn`

**Comment by arvidn** [comment]
oh, it's not a container?

**Comment by justinkb** [comment]
> oh, it's not a container?

I'm using https://github.com/slingamn/namespaced-openvpn to isolate the openvpn tunnel interface from my main network namespace, I will post the ifconfig and netstat outputs tho, still should be useful

**Comment by arvidn** [comment]
libtorrent relies on being able to enumerate network interfaces and the routing table via a `PF_ROUTE`/`NETLINK_ROUTE` socket

**Comment by justinkb** [comment]
Ok, learned something after a reboot. Now that the DHT network has been bootstrapped (when I started the process interactively), I can download torrents by getting peers via DHT, even when having started it via systemd. Neither UDP nor HTTP trackers work now. Does that help you diagnosing the issue? Before I go to sleep for the night, here are those outputs you requested

```
$ sudo ip netns exec protected sudo -u unprivilegeduser ifconfig
lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        inet6 ::1  prefixlen 128  scopeid 0x10<host>
        loop  txqueuelen 1000  (Local Loopback)
        RX packets 1619  bytes 1767643 (1.6 MiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 1619  bytes 1767643 (1.6 MiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

tun0: flags=4305<UP,POINTOPOINT,RUNNING,NOARP,MULTICAST>  mtu 1500
        inet 10.7.3.4  netmask 255.255.255.255  destination 10.7.3.1
        inet6 fe80::5644:7d7a:6313:a721  prefixlen 64  scopeid 0x20<link>
        unspec 00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00  txqueuelen 100  (UNSPEC)
        RX packets 732545  bytes 1011939165 (965.0 MiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 560451  bytes 29701999 (28.3 MiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
```

```
$ sudo ip netns exec protected sudo -u unprivilegeduser ip route
default dev tun0 scope link
10.7.3.1 dev tun0 proto kernel scope link src 10.7.3.4
```

```
sudo ip netns exec protected sudo -u unprivilegeduser netstat -rn
Kernel IP routing table
Destination     Gateway         Genmask         Flags   MSS Window  irtt Iface
0.0.0.0         0.0.0.0         0.0.0.0         U         0 0          0 tun0
10.7.3.1        0.0.0.0         255.255.255.255 UH        0 0          0 tun0
```

Starting to think libtorrent isn't even at fault here and that systemd doesn't handle NetworkNamespacePath specification correctly

**Comment by arvidn** [comment]
thanks! I will look into this

**Comment by justinkb** [comment]
Looks like systemd NetworkNamespacePath specification was indeed the problem, I've worked around it by just using ip netns exec in the ExecStart. I've got everything working perfectly now on libtorrent 1.2.6 and qBittorrent 4.2.5, so this issue can be closed I think. Sorry for bothering you with this ;-(

**Comment by arvidn** [comment]
I think this comment is relevant https://github.com/arvidn/libtorrent/issues/4803#issuecomment-659254971
there are fixes in the works
